masonry/src/widget/widget_ref.rs

```diff
@@ -175,27 +168,36 @@ impl<'w> WidgetRef<'w, dyn Widget> {
     ///
     /// **pos** - the position in local coordinates (zero being the top-left of the
     /// inner widget).
-    pub fn find_widget_at_pos(&self, pos: Point) -> Option<WidgetRef<'w, dyn Widget>> {
-        let mut innermost_widget: WidgetRef<'w, dyn Widget> = *self;
-
-        if !self.state().layout_rect().contains(pos) {
+    pub fn find_widget_at_pos<'a>(
+        &self,
+        root: &'a RenderRoot,
+        pos: Point,
+    ) -> Option<WidgetRef<'a, dyn Widget>> {
+        // Get self from the widget arena to bind it to the arena's lifetime. Is there a way around
+        // this? Also see the comment inside the loop rebinding child to the arena's lifetime.
+        let mut innermost_widget = root.get_widget(self.id()).unwrap();
+
+        if !self.state().window_layout_rect().contains(pos) {
             return None;
         }
 
-        // TODO - Rewrite more elegantly
         loop {
             if let Some(clip) = innermost_widget.state().clip {
-                let relative_pos = pos.to_vec2() - innermost_widget.state().window_origin.to_vec2();
-                // If the widget has a clip, the point must be inside
-                // else we don't iterate over children.
-                if !clip.contains(relative_pos.to_point()) {
+                // If the widget has a clip, the point must be inside, else we don't iterate over
+                // children.
+                let pos = pos - innermost_widget.state().window_origin().to_vec2();
+                if !clip.contains(pos) {
                     break;
                 }
             }
```

Comment:
This part should probably be in `Widget::get_child_at_pos`.

Reply:
Done, `Widget::get_child_at_pos` now checks the clip path of children.

Because `WidgetRef::find_widget_at_pos` does explicitly check the layout rect of `self` before recursing down the tree (else it returns None), I've added a check of the clip path of `self`, too.